diff -Naur libreswan-3.25-orig/programs/pluto/ikev1_main.c libreswan-3.25/programs/pluto/ikev1_main.c
--- libreswan-3.25-orig/programs/pluto/ikev1_main.c	2019-05-07 15:41:26.366181612 -0400
+++ libreswan-3.25/programs/pluto/ikev1_main.c	2019-05-07 15:45:41.780183599 -0400
@@ -2727,30 +2727,11 @@
 					 * Useful if the other peer is
 					 * rebooting.
 					 */
-					if (dst->st_event != NULL &&
-					    dst->st_event->ev_type ==
-						  EVENT_SA_REPLACE &&
-					    !monobefore(monotimesum(mononow(),
-						  deltatime(DELETE_SA_DELAY)),
-						dst->st_event->ev_time)) {
-						/*
-						 * Patch from Angus Lees to
-						 * ignore retransmitted
-						 * Delete SA.
-						 */
-						loglog(RC_LOG_SERIOUS,
-							"received Delete SA payload: already replacing IPSEC State #%lu in %ld seconds",
-							dst->st_serialno,
-							(long)deltasecs(monotimediff(
-								dst->st_event->ev_time,
-								mononow())));
-					} else {
-						loglog(RC_LOG_SERIOUS,
-							"received Delete SA payload: replace IPSEC State #%lu now",
-							dst->st_serialno);
-						dst->st_margin = deltatime(0);
-						event_force(EVENT_SA_REPLACE, dst);
-					}
+					loglog(RC_LOG_SERIOUS,
+						"received Delete SA payload: replace IPSEC State #%lu now",
+						dst->st_serialno);
+					dst->st_margin = deltatime(0);
+					event_force(EVENT_SA_REPLACE, dst);
 				} else {
 					loglog(RC_LOG_SERIOUS,
 						"received Delete SA(0x%08" PRIx32 ") payload: deleting IPSEC State #%lu",
