diff -Naur libreswan-3.23-orig/programs/addconn/addconn.c libreswan-3.23/programs/addconn/addconn.c
--- libreswan-3.23-orig/programs/addconn/addconn.c	2018-01-25 15:19:46.000000000 -0500
+++ libreswan-3.23/programs/addconn/addconn.c	2018-02-05 14:13:25.758711788 -0500
@@ -140,7 +140,9 @@
 	rc |= S_RULE_ADD(set_robust_list);
 	rc |= S_RULE_ADD(set_tid_address);
 	rc |= S_RULE_ADD(socket);
+	rc |= S_RULE_ADD(socketcall);
 	rc |= S_RULE_ADD(socketpair);
+	rc |= S_RULE_ADD(stat);
 	rc |= S_RULE_ADD(statfs);
 	rc |= S_RULE_ADD(uname);
 	rc |= S_RULE_ADD(write);
diff -Naur libreswan-3.23-orig/programs/pluto/pluto_seccomp.c libreswan-3.23/programs/pluto/pluto_seccomp.c
--- libreswan-3.23-orig/programs/pluto/pluto_seccomp.c	2018-01-25 15:19:46.000000000 -0500
+++ libreswan-3.23/programs/pluto/pluto_seccomp.c	2018-02-05 14:13:38.390239502 -0500
@@ -109,6 +109,7 @@
 		rc |= S_RULE_ADD(set_robust_list);
 		rc |= S_RULE_ADD(setsockopt);
 		rc |= S_RULE_ADD(socket);
+		rc |= S_RULE_ADD(socketcall);
 		rc |= S_RULE_ADD(socketpair);
 		rc |= S_RULE_ADD(sysinfo);
 		rc |= S_RULE_ADD(uname);
